Add Float16 to stdlib

stephentyrone opened the pull request:
Implement IEEE 754 binary16 as Float16 in the stdlib conforming to BinaryFloatingPoint and SIMDScalar.

Evolution proposal: https://github.com/apple/swift-evolution/blob/master/proposals/0277-float16.md
SE thread: https://forums.swift.org/t/se-0277-float16/33546

stephentyrone committed:
Implement Float16

stephentyrone committed:
Add support for simd vectors of Float16, and Mirrors.

stephentyrone committed:
Add conversions from Float16 to integer types.

stephentyrone committed:
Quick implementation of FloatingPointParsing for Float16

This isn't quite right, because it introduces double-rounding, but it will do in the short-term.

stephentyrone committed:
Add codable conformance for Float16, remove AnyHashableBox.

stephentyrone committed:
Add importer support for mapping __fp16 and _Float16 to Float16.

stephentyrone committed:
add _swift_stdlib_strtof16 shim

stephentyrone committed:
More float16 work.

stephentyrone committed:
Add 9999 availability to all Float16 API.

stephentyrone commented:
@swift-ci please test

stephentyrone commented:
@swift-ci please test source compatibility

stephentyrone committed:
Add availability to floatingPointConversion validation tests.

stephentyrone commented:
@swift-ci please test

stephentyrone committed:
Adjust diagnostics test so it works with or without Float16.

stephentyrone commented:
@swift-ci please test

lorentey reviewed (approved):
Looks good! I caught one typo 🎣

stephentyrone committed:
Update stdlib/public/core/FloatingPointTypes.swift.gyb

Co-Authored-By: Karoy Lorentey <klorentey@apple.com>

stephentyrone commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -58,6 +67,7 @@ else:
 % end
 
 ${SelfDocComment}
+${Availability(bits)}
```
I don't *think* so.

stephentyrone commented:
Update: ubuntu seems to be missing the float <-> float16 conversion functions in its libgcc, so we'll probably have to build them into the Swift runtime lib on Linux (and likely some other platforms, as well). I'll add a patch to do that sometime in the next few days.

stephentyrone commented:
@swift-ci please smoke test Linux

Local build isn't working right, so rebuilding to get a list of exactly the symbols required.

stephentyrone commented:
@swift-ci please smoke test Linux

stephentyrone commented:
(Local build for Linux isn't working quite right, so rebuilding on CI to get a list of exactly the symbols required.)

stephentyrone commented:
Looks like we need to provide:

- `__gnu_h2f_ieee` (half -> float)
- `__gnu_f2h_ieee` (float -> half)
- `__truncdfhf2` (double -> half)

stephentyrone committed:
Add float16 builtins to the Swift runtime on non-Darwin.

stephentyrone commented:
@swift-ci smoke test

stephentyrone commented:
@tbkka can you review the new runtime/Float16Support.cpp?

tbkka commented on stdlib/public/core/Codable.swift:
```diff
@@ -4592,6 +4592,29 @@ extension RawRepresentable where RawValue == Float, Self: Decodable {
+  ///
+  /// - Parameter decoder: The decoder to read data from.
+  public init(from decoder: Decoder) throws {
+    self = try Self(Float(from: decoder))
```
I'd prefer to be more explicit here about the rounding/truncation behavior.  What happens if someone writes a Float and reads a Float16?

stephentyrone replied:
I think we should do the same thing as Double -> Float does. I'm not super happy with those semantics, but doing anything different is silly. (So: round in-range values, throw if a finite value is out-of-range)

stephentyrone replied:
This is done.

tbkka commented on stdlib/public/runtime/Float16Support.cpp:
```diff
@@ -0,0 +1,164 @@
+//===------------- Float16Support.cpp - Swift Float16 Support -------------===//
```
This seems like the right place for this.

tbkka commented on stdlib/public/core/FloatingPointTypes.swift.gyb:
```diff
@@ -76,13 +86,19 @@ public struct ${Self} {
     } else {
+% if bits == 16:
+      // TODO: replace with real _float16ToString implementation
+      var (buffer, length) = _float32ToString(Float(self), debug: false)
```
Rather than having three copies of this conditional here, feel free to add a new stub for this to stdlib/public/stubs/Stubs.cpp and assign a Radar to me to fill in the SwiftDtoa function.
```
SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
uint64_t swift_float16ToString(char *Buffer, size_t BufferLength,
                               float Value, bool Debug) {
  // FIXME: dispatch to float for now, fill in float16 soon
  return swift_format_float(Value, Buffer, BufferLength);
}
```

stephentyrone replied:
Yeah, this turns out to be quite a bit nicer. I'll do it this way.

tbkka reviewed (approved):
Minor suggestions.

stephentyrone commented:
Getting back to this, let's run a test on Linux and see where we stand ...

stephentyrone commented:
@swift-ci test

stephentyrone committed:
Incorporate suggestions from review.

xwu commented:
@swift-ci test

stephentyrone committed:
Further review simplifications.

Add an actual float16ToString operation, that we can give a real implemention for down the road.

stephentyrone commented:
@swift-ci test

CodaFi commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -995,6 +995,8 @@ struct IEEESemantics {
 
 IEEESemantics getFPSemantics(BuiltinFloatType *fpType) {
   switch (fpType->getFPKind()) {
+  case BuiltinFloatType::IEEE16:
```
Could you remove the `default` below and expand things out here too? (Not that I anticipate us supporting more exotic FP formats than Power).

stephentyrone replied:
Why? We could certainly do it, but it's not clear to me how that benefits us.

CodaFi replied:
I just hit a bizarre crash in the middle of building a client project that tried to enter constant prop for Float16, found this code, and realized that the default provides zero signal for future authors that need to go update the supported formats - like you!

Serendipity leads to nits! 😅

stephentyrone replied:
Can you sketch out what the change you'd like to see actually is? It's not totally obvious to me how you want to improve this.

CodaFi replied:
```
IEEESemantics getFPSemantics(BuiltinFloatType *fpType) {
  switch (fpType->getFPKind()) {
  case swift::BuiltinFloatType::IEEE16:
    return IEEESemantics(16, 5, 10, false);
  case BuiltinFloatType::IEEE32:
    return IEEESemantics(32, 8, 23, false);
  case BuiltinFloatType::IEEE64:
    return IEEESemantics(64, 11, 52, false);
  case BuiltinFloatType::IEEE80:
    return IEEESemantics(80, 15, 63, true);
  case swift::BuiltinFloatType::IEEE128:
    return IEEESemantics(128, 15, 112, false);
  case swift::BuiltinFloatType::PPC128:
    llvm_unreachable("128-bit PowerPC floats are not yet supported.");
  }
  llvm_unreachable("Unexpected FP semantics");
}
```

stephentyrone committed:
Fix suggested by review.

stephentyrone commented:
@swift-ci test

stephentyrone committed:
compiler-rt funcs need to take unsigned short for float16, not short.

stephentyrone commented:
@swift-ci please test

stephentyrone commented:
@swift-ci please test

compnerd commented:
This caused a regression on the android builds :-(

Logs:
https://dev.azure.com/compnerd/3133d6ab-80a8-4996-ac4f-03df25cd3224/_apis/build/builds/29728/logs/289

The new Float16 support requires runtime support from the compiler runtime.  In particular, the AEABI provides the runtime functions `__aeabi_d2h` to truncate a double to a half precision floating point which is missing in the compiler runtime.  I think that the most expedient thing will be to provide an implementation in the standard library or in the shims :-(.  Thoughts @stephentyrone?

stephentyrone commented:
@compnerd Yes, you'll need to update stdlib/public/runtime/Float16Support.cpp to expose `__truncdfhf2` under that name when building for Android. Host platforms' compiler-rt or libgcc are supposed to provide this operation, but they clearly don't.

compnerd commented:
Yeah, why would the system provide the required symbols?  *sigh*  On the bright side, at least it is supposed to be there in future versions.  I'll see if I can get the shims in place for ARMv7.
